azuredevops.go

```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+	if !ok {
+		return
+	}
 
 	client := wharfapi.Client{
 		ApiUrl:     os.Getenv("WHARF_API_URL"),
 		AuthHeader: c.GetHeader("Authorization"),
 	}
 
 	var resp wharfapi.ProjectRunResponse
-	resp, err2 := client.PostProjectRun(
+	resp, err := client.PostProjectRun(
 		wharfapi.ProjectRun{
-			ProjectID:   uint(projectID),
-			Stage:       "prcreatedd",
+			ProjectID:   projectID,
+			Stage:       "prcreated",
 			Branch:      strings.TrimPrefix(t.Resource.SourceRefName, "refs/heads/"),
 			Environment: environment,
 		},
 	)
 
-	if err2 != nil {
-		fmt.Println("Unable to send trigger to wharf-client: ", err2)
-		c.JSON(http.StatusBadRequest, fmt.Sprintf("Unable to send trigger to wharf-client, error: %v", err2))
+	if err != nil {
+		fmt.Println("Unable to send trigger to wharf-client: ", err)
+		err = fmt.Errorf("unable to send trigger to wharf-client: %v", err)
+		ginutil.WriteTriggerError(c, err, "Unable to send trigger to Wharf client.")
 		return
 	}
 
 	c.JSON(http.StatusOK, resp)
 }
 
-func getGitURL(provider wharfapi.Provider, group string, project azureDevOpsProject) (string, error) {
-	providerURL, err := url.Parse(provider.URL)
-
+func bindImportDetails(c *gin.Context) (importBody, error) {
+	i := importBody{}
+	err := c.BindJSON(&i)
 	if err != nil {
-		fmt.Println("Unable to parse provider url: ", err)
-		return "", err
+		return importBody{}, err
 	}
 
-	gitURL := fmt.Sprintf("git@%v:22/%v/%v/_git/%v", providerURL.Host, group, project.Name, project.Name)
-
-	return gitURL, nil
-
+	return i, nil
 }
 
-func getBodyFromRequest(user string, token string, url string) ([]byte, error) {
-	req, err := http.NewRequest(http.MethodGet, url, nil)
+func putProjectWritesError(c *gin.Context, client wharfapi.Client, provider wharfapi.Provider,
+	i importBody, project azureDevOpsProject) (wharfapi.Project, error) {
+	buildDefinitionStr, err := getBuildDefinitionWritesError(c, i, project.Name)
 	if err != nil {
-		fmt.Println("Unable to get: ", err)
-		return []byte{}, err
+		return wharfapi.Project{}, err
 	}
 
-	req.SetBasicAuth(user, token)
-
-	resp, err := http.DefaultClient.Do(req)
+	gitURL, err := getGitURL(provider, i.Group, project)
 	if err != nil {
-		fmt.Println("Unable to do request: ", err)
-		return []byte{}, err
+		fmt.Println("Unable to construct git url ", err)
+		ginutil.WriteComposingProviderDataError(c, err,
+			fmt.Sprintf("Unable to construct git url for project '%s' in group '%s'", project.Name, i.Group))
+		return wharfapi.Project{}, err
 	}
 
-	defer resp.Body.Close()
-
-	if resp.StatusCode != http.StatusOK {
-		fmt.Println("Unable to get. Status code: ", resp.StatusCode)
-		return []byte{}, err
-	}
+	projectInDb, err := client.PutProject(
+		wharfapi.Project{
+			Name:            project.Name,
+			TokenID:         i.TokenID,
+			GroupName:       i.Group,
+			BuildDefinition: buildDefinitionStr,
+			Description:     project.Description,
+			ProviderID:      provider.ProviderID,
+			GitURL:          gitURL})
```

Comment:
nit: Suggest placing the `wharfapi.Project{` line on the same line as `client.PutProject` to reduce indentation.

```suggestion
	projectInDb, err := client.PutProject(wharfapi.Project{
		Name:            project.Name,
		TokenID:         i.TokenID,
		GroupName:       i.Group,
		BuildDefinition: buildDefinitionStr,
		Description:     project.Description,
		ProviderID:      provider.ProviderID,
		GitURL:          gitURL})
```